src/main/kotlin/lgp/core/evolution/ResultsExporter.kt

```diff
@@ -0,0 +1,339 @@
+     * between threads when results are being added/output simultaneously. Despite this, there is no
+     * guarantee that the order of the results will be sequential in a multi-threaded environment. The
+     * built-in [Trainers] ensure that the results will be grouped by their run however.
+     *
+     * @param outputProvider A [ResultOutputProviders.CsvResultOutputProvider] instance.
+     * @param bufferSize How many results to keep in memory before writing to [outputProvider].
+     * @param verbose Writes detailed output about any buffering operations to standard out.
+     */
+    class BufferedResultAggregator<T>(
+            private val outputProvider: ResultOutputProviders.CsvResultOutputProvider<T>,
+            private val bufferSize: Int = 100,
+            private val verbose: Boolean = false
+    ) : ResultAggregator<T>() {
+
+        override val results: MutableList<ExportableResult<T>> = mutableListOf()
+
+        /**
+         * Adds a single [ExportableResult].
+         */
+        override fun add(result: ExportableResult<T>) {
+            this.maybeFlushBuffer(1)
+
+            synchronized(this) {
+                this.results.add(result)
+            }
+        }
+
+        /**
+         * Adds a collection of [ExportableResult] instances.
+         */
+        override fun addAll(results: List<ExportableResult<T>>) {
+            this.maybeFlushBuffer(results.size)
+
+            this.log("Adding ${results.size} results...")
+            if (results.size > this.bufferSize) {
+                this.addInChunks(results)
+                return
+            }
+
+            synchronized(this) {
+                this.results.addAll(results)
+            }
+        }
+
+        /**
+         * Internal method used to write a collection of [ExportableResult] instances in chunks.
+         *
+         * Writing in chunks will be more efficient than over-filling the buffer and having to flush
+         * a large amount of results in one go.
+         */
+        private fun addInChunks(results: List<ExportableResult<T>>) {
+            this.log("Adding in chunks...")
+
+            results.chunked(this.bufferSize).map { chunk ->
+                this.log("Adding chunk with ${chunk.size} results...")
+                this.addAll(chunk)
+            }
+        }
+
+        /**
+         * Writes any remaining buffered results.
+         */
+        override fun close() {
+            this.log("Writing any remaining buffered results...")
+            if (results.isEmpty())
+                return
+
+            this.outputProvider.writeResultsFrom(this)
+            
+            synchronized(this) {
+                this.results.clear()
+            }
+        }
+
+        /**
+         * Determines if the buffer needs flushing and flushes if necessary.
+         */
+        private fun maybeFlushBuffer(count: Int) {
+            when {
+                (results.size + count) < this.bufferSize -> return
```

Comment:
Describe why return when this is true. Yes it is clear from the logic, but good to be explicit.